### PR TITLE
build(devtools): migrate to manifest v3

### LINKED ISSUE #46287: Update Angular DevTools manifest version
### Which @angular/* package(s) are relevant/related to the feature request?

_No response_

### Description

We currently use Manifest version 2, which is deprecated and Chrome will no longer support it starting in 2023. [Read more](https://developer.chrome.com/blog/mv2-transition/).

### Proposed solution

Explore the v3 format and update.

### Alternatives considered

None.

### PR DESCRIPTION
**Pull request by AleksanderBodurri**
Previously we built DevTools for all browsers with version 2 of the manifest file format.

This commit includes a number of refactors and API additions that will enable us to build DevTools with version 3 of the manifest file format.

The manifest v3 build of Angular DevTools has been tested on Chrome, Edge, and Safari.

Notably, the Firefox version of Angular DevTools remains as a manifest v2 build. Firefox does not yet support manifest v3 in it's latest stable release. When Firefox makes this transition, a follow up PR will update the Firefox manifest file to version 3.

Because Firefox still needs v2, we need to keep some old v2 APIs around in our background page (service worker in v3) that will execute conditionally based on if the extension was built for v2 or v3. This is determined with the `chrome.runtime.getManifest().manifest_version` API.

# Changes

> `ng_validate.ts` content script was refactored into `ng_validate.ts` and `detect-angular-for-extension-icon.ts`. Previously this script would execute code on the main page that would detect the existence of Angular and communicate it to the DevTools background page in order to update the Angular icon and popup in the toolbar.
>
> This code would execute on the main page through a script tag that we supply a function to like so, `script.textContent = functionAsString`. 
>
> With v3, this method of executing code in a content script [no longer works](https://developer.chrome.com/docs/extensions/mv3/intro/mv3-migration/#code-execution), so we follow the [migration guide](https://developer.chrome.com/docs/extensions/mv3/mv3-migration-checklist/#api-remote-code) to [resolve this](https://github.com/angular/angular/blob/63f158cdf5012d577187dba2a404cec416084eca/devtools/projects/shell-browser/src/app/ng-validate.ts#L18). 
>
> The only notable change here is that our detection code no longer executes synchronously with `ng_validate.ts`. This is not an issue as this content script was always loaded asynchronously in the first place, and we have no such feature in Angular DevTools that requires it to execute synchronously.

> The `chrome.tabs.executeScript` API no longer works in v3. As [per the guide](https://developer.chrome.com/docs/extensions/mv3/mv3-migration-checklist/#api-tabs), we switch to using `chrome.scripting.executeScript`.

> The `chrome.browserAction` API was previously used to set the icon and popup in the browser toolbar. In v3, this API along with the `chrome.pageAction` API was merged into the `chrome.action` API. As [per the guide](https://developer.chrome.com/docs/extensions/mv3/mv3-migration-checklist/#api-browser-action-js), we make this change.

>[This guide](https://developer.chrome.com/docs/extensions/mv3/intro/mv3-migration/#updating-manifest-dot-json) to migrating the manifest.json file was followed.

closes #46287

### CODE REVIEW

File: devtools/projects/ng-devtools/src/lib/devtools.component.ts [hunk 1]
```diff
@@ -72,7 +72,7 @@ export class DevToolsComponent implements OnInit, OnDestroy {
   }
 
   get supportedVersion(): boolean {
-    return (this.majorAngularVersion >= 9 || this.majorAngularVersion === 0) && this.ivy;
+    return (this.majorAngularVersion >= 12 || this.majorAngularVersion === 0) && this.ivy;
```
Comment (mgechev) [inline devtools/projects/ng-devtools/src/lib/devtools.component.ts:75]: This seems like an unrelated change. Can we move to a different commit?

File: devtools/projects/shell-browser/src/app/background.ts [hunk 2]
```diff
@@ -8,19 +8,29 @@
...
+})();
 
 // By default use the black and white icon.
 // Replace it only when we detect an Angular app.
 browserAction.setIcon({
   path: {
-    16: `assets/icon-bw16.png`,
-    48: `assets/icon-bw48.png`,
-    128: `assets/icon-bw128.png`,
+    16: chrome.runtime.getURL(`assets/icon-bw16.png`),
```
Comment (mgechev) [inline devtools/projects/shell-browser/src/app/background.ts:31]: `chrome.runtime.getURL` is valid for both v2 and v3?

File: devtools/projects/shell-browser/src/app/background.ts [hunk 3]
```diff
@@ -153,19 +174,23 @@ chrome.runtime.onMessage.addListener((req, sender) => {
   if (!req.isAngularDevTools) {
     return;
   }
-  if (sender && sender.tab) {
-    browserAction.setPopup({
-      tabId: sender.tab.id,
-      popup: `popups/${getPopUpName(req)}`,
-    });
+
+  if (!sender?.tab) {
```
Comment (mgechev) [inline devtools/projects/shell-browser/src/app/background.ts:178]: This seems like a bug fix? We're now going to show the popup even if the app is not built with Angular?

File: devtools/projects/shell-browser/src/app/detect-angular-for-extension-icon.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  const ngVersionElement = document.querySelector('[ng-version]');
+  let isSupportedAngularVersion = false;
+  let isAngular = false;
+
+  if (ngVersionElement) {
+    isAngular = true;
+    const attr = ngVersionElement.getAttribute('ng-version');
+    const major = attr ? parseInt(attr.split('.')[0], 10) : -1;
+    // In case of g3 apps we support major 0.
+    if (attr && (major >= 12 || major === 0)) {
```
Comment (mgechev) [inline devtools/projects/shell-browser/src/app/detect-angular-for-extension-icon.ts:33]: With a subsequent PR we may want to use `const SupportedMajorVersion = 12`.

**Review by mgechev — APPROVED** [review]
LGTM with one small nit!

File: devtools/projects/shell-browser/src/app/background.ts (hunk 3 above)
Comment (AleksanderBodurri) [reply devtools/projects/shell-browser/src/app/background.ts]: The logic before and after this change is identical, although a part of the refactor here is unnecessary for this PR. I'll change it back to keep the changes more focused to manifest v3 API changes.

File: devtools/projects/shell-browser/src/app/background.ts (hunk 2 above)
Comment (AleksanderBodurri) [reply devtools/projects/shell-browser/src/app/background.ts:31]: Yes, this API has been around since Chrome 22. https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/getURL#browser_compatibility

**Review by mgechev — APPROVED** [review]

**Comment by jessicajaniuk** [comment]
@AleksanderBodurri CI is broken here. Can you rebase?

**Review by jessicajaniuk — APPROVED** [review]

**Comment by jessicajaniuk** [comment]
This PR was merged into the repository by commit a910c8328aa6d28ac147141babbe166c19439519.
